httpclient_android/http_client/src/main/java/com/backblaze/b2/client/webApiHttpClient/android/UrlConnection/B2WebApiHttpClientImpl.java

```diff
@@ -0,0 +1,414 @@
+        }
+    }
+
+    private B2Exception translateToB2Exception(IOException e, String url) {
+        if (e instanceof ConnectException) {
+            // java.net base class for HttpHostConnectException.
+            return new B2ConnectFailedException("connect_failed", null, "failed to connect for " + url, e);
+        }
+        if (e instanceof UnknownHostException) {
+            return new B2ConnectFailedException("unknown_host", null, "unknown host for " + url, e);
+        }
+        if (e instanceof SocketTimeoutException) {
+            return new B2NetworkTimeoutException("socket_timeout", null, "socket timed out talking to " + url, e);
+        }
+        if (e instanceof SocketException) {
+            return new B2NetworkException("socket_exception", null, "socket exception talking to " + url, e);
+        }
+        if (e instanceof MalformedURLException) {
+            return new B2NotFoundException("malformed_url", null, "malformed for " + url, e);
+        }
+
+        return new B2NetworkException("io_exception", null, e + " talking to " + url, e);
+    }
+
+    private B2Exception extractExceptionFromErrorResponse(Class<?> response,
+                                                          String responseText) {
+        final int statusCode = response.getResponseCode();;
+        // Try B2 error structure
+        try {
+            final B2ErrorStructure err = B2Json.get().fromJson(responseText, B2ErrorStructure.class);
+            return B2Exception.create(err.code, err.status, getRetryAfterSecondsOrNull(response), err.message);
+        }
+        catch (Throwable t) {
+            // we can't parse the response as a B2 JSON error structure.
+            // so use the default.
+            return new B2Exception("unknown", statusCode, getRetryAfterSecondsOrNull(response), responseText);
+        }
+    }
+
+    /**
+     * If there's a Retry-After header and it has a delay-seconds formatted value,
+     * this returns it.  (to be clear, if there's an HTTP-date value, we ignore it
+     * and keep looking for one with delay-seconds format.)
+     *
+     * @param response the http response.
+     * @return the delay-seconds from a Retry-After header, if any.  otherwise, null.
+     */
+    private Integer getRetryAfterSecondsOrNull(URLConnection response) {
+        // https://tools.ietf.org/html/rfc7231#section-7.1.3
+        for (String header : response.getHeaderFields().get(B2Headers.RETRY_AFTER)) {
+            try {
+                return Integer.parseInt(header.getValue(), 10);
+            } catch (IllegalArgumentException e) {
+                // continue.
+            }
+        }
+
+        return null;
+    }
+
+    private HashMap<String, String> makeHeaders(B2Headers headersOrNull) {
+        if (headersOrNull == null) {
+            return new HashMap<>();
+        }
+        final HashMap<String, String> headers = new HashMap<>();
+
+        for (String name : headersOrNull.getNames()) {
```

Comment:
it is legal to have multiple headers with the same name, but this conversion will eliminate all but the last one.